### PR TITLE
[LAD] IMDS probe feature informational PR

### PR DESCRIPTION
**Pull request by hosungsmsft**
For Abhi and Hari

### CODE REVIEW

File: Diagnostic/Utils/imds_util.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            imds_data = '{0}'.format(e)
+
+        msg = 'IMDS instance data = {0}'.format(imds_data)
+        if log_as_ext_event:
+            self._ext_event_logger(name=self._ext_name,
+                                   op=self._ext_op_type,
+                                   isSuccess=True,
+                                   version=self._ext_ver,
+                                   message=msg)
+        if self._ext_logger:
```
Comment (jasonzio) [inline Diagnostic/Utils/imds_util.py]: Refactor this two-line "test and call" in a single always-valid method.

File: Diagnostic/Utils/imds_util.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+import datetime
+import urllib2
+import time
+import traceback
+
+
+def get_imds_data(node, json=True):
+    """
+    Query IMDS endpoint for instance metadata and return the response as a Json string.
```
Comment (hosungsmsft) [inline Diagnostic/Utils/imds_util.py]: Done

File: Diagnostic/Utils/imds_util.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    """
+    if not node:
+        return None
+    if node[0] != '/':
```
Comment (hosungsmsft) [inline Diagnostic/Utils/imds_util.py]: Done

File: Diagnostic/Utils/imds_util.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    """
+
+    def __init__(self, ext_name, ext_ver, ext_op_type, ext_event_logger, ext_logger=None,
+                 imds_data_getter=get_imds_data, logging_interval_in_minutes=60):
+        self._ext_name = ext_name
+        self._ext_ver = ext_ver
+        self._ext_op_type = ext_op_type
+        self._ext_logger = ext_logger  # E.g., hutil.log
+        self._ext_event_logger = ext_event_logger  # E.g., waagent.AddExtensionEvent
+        self._last_log_time = None
```
Comment (hosungsmsft) [inline Diagnostic/Utils/imds_util.py]: Done

File: Diagnostic/Utils/imds_util.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        self._imds_data_getter = imds_data_getter
+        self._logging_interval = datetime.timedelta(minutes=logging_interval_in_minutes)
+
+    def log_imds_data_if_right_time(self, log_as_ext_event=False):
+        now = datetime.datetime.now()
+        not_yet = self._last_log_time and now < self._last_log_time + self._logging_interval
```
Comment (hosungsmsft) [inline Diagnostic/Utils/imds_util.py]: Done

File: Diagnostic/Utils/imds_util.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        self._ext_ver = ext_ver
+        self._ext_op_type = ext_op_type
+        self._ext_logger = ext_logger  # E.g., hutil.log
+        self._ext_event_logger = ext_event_logger  # E.g., waagent.AddExtensionEvent
+        self._last_log_time = None
+        self._imds_data_getter = imds_data_getter
+        self._logging_interval = datetime.timedelta(minutes=logging_interval_in_minutes)
+
+    def log_imds_data_if_right_time(self, log_as_ext_event=False):
+        now = datetime.datetime.now()
```
Comment (hosungsmsft) [inline Diagnostic/Utils/imds_util.py]: Done

File: Diagnostic/Utils/imds_util.py (hunk 1 above)
Comment (hosungsmsft) [inline Diagnostic/Utils/imds_util.py]: Good idea. Thanks for all comments. Done.

**Pull request merged**